test/extensions/transport_sockets/tls/cert_validator/default_validator_integration_test.cc

```diff
@@ -64,8 +64,8 @@ TEST_P(SslCertValidatorIntegrationTest, CertValidated) {
   codec->close();
 }
 
-// With verify-depth set, certificate validation fails
-TEST_P(SslCertValidatorIntegrationTest, CertValidationFailed) {
+// With verify-depth set, certificate validation succeeds since we allow partial chain
```

Comment:
It looks like setting this PR breaks the existing feature for limiting verify depth. Is that correct?

Reply:
Or is this test configured to trust all the intermediate CAs, and so the end of the max_verify_depth points to a trusted intermediate?

If that's the case, please add another test case with not-trusted intermediates and a max_verify_depth and validate that it fails.

Reply:
max_verify_depth limits the certificate chain size when constructing chain during verification.

Assuming there is a full certificate chain like this: cert -> intermediate 2 ca < intermediate 1 ca < root ca.
If max_verify_depth is set to 2 , the chain constructed will be cert -> intermediate 2 ca < intermediate 1 ca
with X509_V_FLAG_PARTIAL_CHAIN, we don't need to trust all intermediate CAs, only trust intermediate 2 ca is enough (intermediate 2 ca is treated as trust anchor).
without X509_V_FLAG_PARTIAL_CHAIN, validation fails because intermediate CAs are not treated as trust anchor.

I can add a test with max_verify_depth set and intermediate 2 ca not in trust_ca, it that you required?

Reply:
> I can add a test with max_verify_depth set and intermediate 2nd ca not in trust_ca, is that what you required?

Yes, please add that test, otherwise there isn't a test case anymore for failing to validate due to max_depth anymore.

Reply:
I need to clarify a little bit before adding this test.

I think the failing for this case you required is not due to max_depth, it 's due to no valid partial chain can be constructed based on trust ca.
That is to say if we don't set max_depth at all, and trust_ca does not contain intermediate 2nd ca, it still fails because we can't construct valid partial chain without trust intermediate 2nd ca.

The description for [max_verify_depth](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext) is not accurate:

> If configured, and the certificate chain is longer than allowed, the certificates above the limit are ignored, and certificate validation will fail.

This is just the old behavior in Envoy, not exact meaning of the max_verify_depth.

We should reword like [openssl doc](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_verify_depth.html):
> Certificate chain must reach a trusted issuer no longer than allowed. Otherwise, certificate validation will fail.

Reply:
> not sure I understand correctly. I'm thinking the case of the user setting a chain which longer than max_depth, but with `X509_V_FLAG_PARTIAL_CHAIN`, it only validates part of them. And that happened silently.
> 
It happened silently.

Reply:
The case I'm interested in here is leaf->intermediate1->intermediate2->root, with max_verify_depth set to 1 (or 2?). If only the root is trusted, this should fail. If the intermediate is trusted, it should successfully verify.

Reply:
> The case I'm interested in here is leaf->intermediate1->intermediate2->root, with max_verify_depth set to 1 (or 2?). If only the root is trusted, this should fail. If the intermediate is trusted, it should successfully verify.

Thus we need to make this as an option instead of force enabling, introducing a field "partial_chain" into [`CertificateValidationContext`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext), and set its default value to true in tests. Then we can choose if we only trust root or trust intermediate. Is it OK?

Reply:
@ggreenway what do you think about adding a config option for `X509_V_FLAG_PARTIAL_CHAIN`, if no and setting `X509_V_FLAG_PARTIAL_CHAIN` silently, the depth can not control anything.

new config option can provide flexibility, silent setting can keep API consistent, not sure one is better.

Reply:
I don't think an option is needed. You just need to set the trusted ca certs appropriately in the test. Two tests, both setting max-verify-depth to a length less than the full chain. In one of the tests, you set the intermediate as trusted. The validation will succeed in this test. In the other test, you only set the root of the chain as trusted, and not the intermediates. In this test, the validation will fail.

Reply:
@ggreenway 

> you only set the root of the chain as trusted, and not the intermediates

Do you mean only root ca in trust_ca file? In this case, validation will always fail, it has nothing to do with max-verify-depth. Let me explain case by case.

trust_ca should contain left partial chain or full chain. For a leaf cert -> intermediate2 -> intermediate1 -> root,

1. trust_ca contains intermediate2 -> intermediate1 or intermediate2 (partial chian)
if we don't set X509_V_FLAG_PARTIAL_CHAIN, verification always fails since only root can be trust anchor
If we set X509_V_FLAG_PARTIAL_CHAIN, verification always succeeds since intermediate ca can be trust anchor

2. trust_ca contains  intermediate2 -> intermediate1 -> root (full chain)
if we don't set X509_V_FLAG_PARTIAL_CHAIN, verification result depends on max-verify-depth, it succeeds when the depth >= root ca depth
If we set X509_V_FLAG_PARTIAL_CHAIN, verification always succeeds since intermediate ca can be trust anchor

4. trust_ca contains intermediate1 -> root or intermediate1 or root (broken chain)
validation always fails, because we can not build chain with out intermediate 2 in trusted ca.

If I understand correctly, it will try to build a chain during velidation, all certs in this chain must be in trust ca otherwise chain building will stop. max-verify-depth also have effects on the length of the chain.
X509_V_FLAG_PARTIAL_CHAIN just to indicate if the intermediate ca can be trust anchor, if not set our chain must contain root ca.

Thus to satisfy your requirement on the test case, we must make the X509_V_FLAG_PARTIAL_CHAIN to be an option.

Reply:
> If I understand correctly, it will try to build a chain during velidation, all certs in this chain must be in trust ca otherwise chain building will stop.

This is incorrect. The chain-building tries to find a chain from the leaf to one of the certs in `trusted_ca`. Without FLAG_PARTIAL_CHAIN, it will only consider self-signed certs when looking for a trusted cert (it treats anything that isn't self-signed as if it weren't in the `trusted_ca` at all). With FLAG_PARTIAL_CHAIN, it will also consider intermediate certs IFF they're in `trusted_ca`. But the intermediates do NOT need to be in `trusted_ca` as long as something in their cert-chain is in `trusted_ca`.

Reply:
>  But the intermediates do NOT need to be in trusted_ca as long as something in their cert-chain is in trusted_ca.

I don't think it's correct.
See
https://github.com/envoyproxy/envoy/blob/5261b4285d1487f93e7aece94786de9a8d3a2529/test/config/integration/certs/certs.sh#L80-L81

https://github.com/envoyproxy/envoy/blob/5261b4285d1487f93e7aece94786de9a8d3a2529/test/config/utility.cc#L1374-L1380

https://github.com/envoyproxy/envoy/blob/18212bb6395af308d895f75352f82df522b038b4/test/extensions/transport_sockets/tls/cert_validator/default_validator_integration_test.cc#L53-L65

The test will fail if we remove any one of the `cacert.pem` `intermediate_cacert.pem`  `intermediate_ca_2cert.pem` from trust ca. This is what I observe from testing (without FLAG_PARTIAL_CHAIN) 

So I infer that
1. If the `intermediate_ca_2cert.pem` is not in trust ca, the built chain will only contains leaf
2. if the `intermediate_cacert.pem` is not in trust ca, the built chain will be leaf -> intermediate_ca_2cert
3. if the `cacert.pem` is not in trust ca, the built chain will be leaf -> intermediate_ca_2cert -> intermediate_cacert.
Without FLAG_PARTIAL_CHAIN, only root ca(self-signed) can be trust anchor, so we have to build a full chain for successful validation.

For the testing with FLAG_PARTIAL_CHAIN, 2 and 3 will succeed, since the intermediate ca can be trust anchor.

Reply:
There's a lot going on here, but it's all related to how this specific test is setup.

* The client is not sending a chain; it's only sending the leaf cert
* Because the client doesn't send the chain, validation will fail on the server unless the server has the full chain. Putting the full chain in the trusted ca accomplishes this.
* The chain is actually malformed here: intermediate_ca is configured with `pathlen:0`, which means that it is not valid for it to sign intermediate_2. This misconfiguration requires that intermediate2 be a trusted CA for validation to succeed. I'm pretty sure that setting `pathlen:0` on intermediate CA is a bug in the test setup, and should be fixed by setting `pathlen:1` both locations in `intermediate_cacert.cfg`.

To test this new flag, I think you'll need to generate the client cert full chain, via `cat client2cert.pem intermediate_ca_2cert.pem intermediate_cacert.pem cacert.pem > client2_chain.pem`. Then you should be able to only add the root CA to trusted CA configuration.

Reply:
@ggreenway 

When I fix the `pathlen` issue, generate the client full chain, and only add root CA to trusted CA configuration, the test still succeeds to validate cert with` max_verify_depth = 1` (FLAG_PARTIAL_CHAIN is set).  It seems no difference that we put chain on client side or server side.

If we put the chain on client cert, which level does it start to calculate depth? Does it still start from leaf cert?

Reply:
I would have assumed it would start from the leaf cert, and would fail in this configuration. Can you post the test code for this?

Reply:
The test code for this is: 
https://github.com/envoyproxy/envoy/blob/765e7ab10b2cd72742a238de6e75648d07a6a985/test/extensions/transport_sockets/tls/cert_validator/default_validator_integration_test.cc#L101

Reply:
@ggreenway 

Dig deeper... It seems boringssl hiding the error, there is a FIXME in `X509_verify_cert` implementation:

https://github.com/google/boringssl/blob/d45d8933e61ccd8bd50fcf58f33a604627e6552c/crypto/x509/x509_vfy.c#L228-L233

Reply:
@LuyaoZhong There is a mistake in your test, the `max_verify_depth` doesn't set currently at https://github.com/envoyproxy/envoy/pull/22350/files#diff-ce04c82dd2b166c6df41e514e7c253ceeb276f7db1d186ddc16f8f1841ad9496R1391

In the end the `max_verify_depth` missed in the downstream transport config.

That is why your test doesn't work.